File: SEP18/5.py
from math import gcd

def fun(h,w, photo, col,row):
	temp = []
	for i in range(h):
		temp.append(photo[i * w: (i * w) + w])
	new_photo = ""
	for i in temp:
		temp1 = ""
		for j in i:
			new_photo += j * col;
			temp1 += j * col;
		new_photo += temp1 * (row - 1)
	# print(new_photo)
	return new_photo
	# print(new_photo)
	# print(len(new_photo))
	# print(temp)

def find_similarity(photo1,photo2):
	count = 0
	for i in range(len(photo1)):
		if photo1[i] == photo2[i]:count += 1
	return count

for _ in range(int(input())):
	h1,w1 = [int(x) for x in input().split()]
	photo1 = input()
	h2,w2 = [int(x) for x in input().split()]
	photo2 = input()
	new_photo_row = h1 * h2 // gcd(h1,h2)
	new_photo_col = w1 * w2 // gcd(w1,w2)
	print(new_photo_row, new_photo_col)
	size_of_pixel_in_new_photo_col = new_photo_col // w1
	size_of_pixel_in_new_photo_row = new_photo_row // h1
	new_photo1 = fun(h1,w1,photo1,size_of_pixel_in_new_photo_col, size_of_pixel_in_new_photo_row)
	size_of_pixel_in_new_photo_col = new_photo_col // w2
	size_of_pixel_in_new_photo_row = new_photo_row // h2
	new_photo2 = fun(h2,w2,photo2,size_of_pixel_in_new_photo_col, size_of_pixel_in_new_photo_row)
	print(find_similarity(new_photo1, new_photo2))
	# fun(h2,w2,photo2)

File: SEP18/5.cpp
#include<iostream>
#include<algorithm>
using namespace std;
string photo1,photo2;

int isPixelSame(long i,long j,int new_pixel_photo1_row, int new_pixel_photo1_col, int new_pixel_photo2_row, int new_pixel_photo2_col, int h1, int w1, int h2, int w2){
	int new_i_photo1 = int(i / new_pixel_photo1_row);
	int new_j_photo1 = int(j / new_pixel_photo1_col);
	int new_i_photo2 = int(i / new_pixel_photo2_row);
	int new_j_photo2 = int(j / new_pixel_photo2_col);
	// cout<<i<<" " << j <<" " << (new_i_photo1 * (w1)) + (new_j_photo1)<<" " << (new_i_photo2 * (w2)) + (new_j_photo2)<<endl;
	return photo1[(new_i_photo1 * (w1)) + (new_j_photo1)] == photo2[(new_i_photo2 * (w2)) + (new_j_photo2)];
}


int main(void){
	std::ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	int t;cin>>t;
	while(t--){
		int h1,w1;cin>>h1>>w1;
		photo1;cin>>photo1;
		int h2,w2;cin>>h2>>w2;
		photo2;cin>>photo2;
		long new_photo_row = long((h1 * h2) / __gcd(h1,h2));
		long new_photo_col = long((w1 * w2) / __gcd(w1,w2));
		int new_size_of_pixel_photo1_row = int(new_photo_row / h1);
		int new_size_of_pixel_photo1_col = int(new_photo_col / w1);
		int new_size_of_pixel_photo2_row = int(new_photo_row / h2);
		int new_size_of_pixel_photo2_col = int(new_photo_col / w2);
		// cout<<h1<<" "<<h2<<endl;
		// cout<<w1<<" "<<w2<<endl;
		// cout<<__gcd(h1,h2)<<endl;
		// cout<<new_photo_row<<" "<<new_photo_col<<endl;
		long count = 0;
		int i = 0, j = 0;
		int a = new_size_of_pixel_photo1_col,b = new_size_of_pixel_photo2_col, c = new_size_of_pixel_photo1_row, d = new_size_of_pixel_photo2_row;
		while(i < new_photo_row){
			while(j < new_photo_col){
				int temp = isPixelSame(i,j,new_size_of_pixel_photo1_row,new_size_of_pixel_photo1_col,new_size_of_pixel_photo2_row,new_size_of_pixel_photo2_col,h1,w1,h2,w2);
				if(temp == 1){
					long m1 = min(a,b);
					long m2 = min(c,d);
					count += m1 + m2;
					j += m1;
					i += m2;
					a -= m1;
					b -= m1;
					c -= m2;
					d -= m2;
					if(a == 0) a = new_size_of_pixel_photo1_col;
					if(b == 0) b = new_size_of_pixel_photo2_col;
					if(c == 0) c = new_size_of_pixel_photo1_row;
					if(d == 0) d = new_size_of_pixel_photo2_row;
				}else{
					j ++;
				}
			}
			i ++;
		}
		// for(long i = 0; i < new_photo_row; i++){
		// 	for(long j = 0; j < new_photo_col; j++){
		// 		count += isPixelSame(i,j,new_size_of_pixel_photo1_row,new_size_of_pixel_photo1_col,new_size_of_pixel_photo2_row,new_size_of_pixel_photo2_col,h1,w1,h2,w2);
		// 		// cout<<i<<" " << j<<" ";
		// 	}
		// 	// cout<<endl;
		// }
		cout<<count<<endl;
	}
}
